storage/remote/azuread/token_provider.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2023 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package azuread
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+)
+
+const (
+	// Clouds
+	AzureChina      = "AzureChina"
+	AzureGovernment = "AzureGovernment"
+	AzurePublic     = "AzurePublic"
+
+	// Audiences
+	INGESTION_CHINA_AUDIENCE      = "https://monitor.azure.cn//.default"
+	INGESTION_GOVERNMENT_AUDIENCE = "https://monitor.azure.us//.default"
+	INGESTION_PUBLIC_AUDIENCE     = "https://monitor.azure.com//.default"
+)
+
+// tokenProvider is used to store and retrieve Azure AD accessToken
+type tokenProvider struct {
+	// token is member used to store the current valid accessToken
+	token string
+	ctx   context.Context
+	// refreshDuration is used to store the refresh time duration of the current valid accessToken
+	refreshDuration time.Duration
+	// credentialClient is the Azure AD credential client that is being used to retrive accessToken
+	credentialClient azcore.TokenCredential
+	options          *policy.TokenRequestOptions
+}
+
+// TokenProvider is the interface for Credential client
+type TokenProvider interface {
+	getAccessToken() string
+}
+
+// newTokenProvider helps to fetch accessToken for different types of credential. This also takes care of
+// refreshing the accessToken before expiry. This accessToken is attached to the Authorization header while making requests.
+func newTokenProvider(ctx context.Context, cfg *AzureADConfig, cred azcore.TokenCredential) (TokenProvider, error) {
+	audience, err := getAudience(cfg.Cloud)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenProvider := &tokenProvider{
+		ctx:              ctx,
+		credentialClient: cred,
+		options:          &policy.TokenRequestOptions{Scopes: []string{audience}},
+	}
+
+	err = tokenProvider.setAccessToken()
+	if err != nil {
+		return nil, errors.New("Failed to get access token: " + err.Error())
+	}
+
+	go tokenProvider.periodicallyRefreshClientToken()
```

Comment:
Instead of having a goroutine do this in the background, could we just check if the token has expired on each request, and if so, update it?

Reply:
The idea behind having a goroutine in the background was that we could completely isolate the consumer of the `accessToken` from the refresh logic. Also based to volume of requests we expect in remote write operations, the overhead having background goroutine vs checking for each request seems comparable. Let me know if you would still like me to change the code to check for every request ?